### PR TITLE
Distributed v3

### PR DESCRIPTION
**Pull request by matthiasdiener**
See also #65 #139.

Needs:
- [x] #139
- [x] #189
- [x] #208

Required by:
- [ ] https://github.com/illinois-ceesd/mirgecom/pull/556
- [ ] https://github.com/illinois-ceesd/mirgecom/pull/562
---

- create functionality to split according to node types as a class DistributedGraphRuntime **or** two functions: prepare+execute
  - remove send, recv nodes for codegen
  - replace recv node with placeholder
  - add send’s predecessor to DictOfNamedArrays

image

### CODE REVIEW

File: pytato/transform.py [hunk 1]
```diff
@@ -1018,12 +1092,17 @@ def find_partitions(expr: Array, part_func: Callable[[Array], PartitionId]) ->\
 
     # codegen
     from pytato import generate_loopy
+
+    dm = DistributedMapper()
```
Comment (inducer) [inline pytato/transform.py:1096]: Codegen should be separate from `find_partition`.

Comment (matthiasdiener) [reply pytato/transform.py:1107]: done in the dependent PR.

File: pytato/array.py [hunk 2]
```diff
@@ -2232,4 +2232,60 @@ def vdot(a: Array, b: Array) -> ArrayOrScalar:
...
+
+# {{{ Distributed execution
+
+import numpy as np
+
+
+# {{{ Communication nodes
+
+
+class DistributedSend(Array):
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/array.py:2271]: done

File: pytato/transform.py [hunk 3]
```diff
@@ -958,10 +991,51 @@ def map_index_lambda(self, expr: IndexLambda, *args: Any) -> IndexLambda:
...
+            -> DistributedRecv:
+        new_binding = self._handle_new_binding(expr, expr.data)
+
+        return DistributedRecv(new_binding)  # FIXME: Return Placeholder instead?
+
     def __call__(self, expr: Array, *args: Any, **kwargs: Any) -> Array:
         return self.rec(expr)
 
 
+class DistributedMapper(CopyMapper):
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/transform.py:992]: done

**Comment by matthiasdiener** [comment]
@inducer, I think this is ready for another look. In particular, it's not clear to me what `post_receives` should actually return (it is more than the MPI request?)

File: pytato/transform.py [hunk 4]
```diff
@@ -226,6 +227,14 @@ def map_reshape(self, expr: Reshape) -> Reshape:
                        order=expr.order,
                        tags=expr.tags)
 
+    def map_distributed_send(self, expr: DistributedSend) -> DistributedSend:
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/transform.py]: Moot now

File: pytato/transform.py [hunk 5]
```diff
@@ -226,6 +227,14 @@ def map_reshape(self, expr: Reshape) -> Reshape:
...
+        return DistributedSend(self.rec(expr.data),
+                               dest_rank=expr.dest_rank, comm_tag=expr.comm_tag)
+
+    def map_distributed_recv(self, expr: DistributedRecv) -> DistributedRecv:
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/transform.py]: Moot now

File: pytato/transform.py [hunk 6]
```diff
@@ -994,16 +1076,102 @@ def find_partitions(expr: Array, part_func: Callable[[Array], Hashable]) ->\
...
+    def __init__(self) -> None:
+        super().__init__()
+
+        from pytools import UniqueNameGenerator
+        self.name_generator = UniqueNameGenerator(forced_prefix="_dist_ph_")
+
+        self.part_input_name_to_recv_node: Dict[str, DistributedRecv] = {}
+        self.part_output_name_to_send_node: Dict[str, DistributedSend] = {}
+
+    def map_distributed_recv(self,  # type: ignore
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/transform.py]: Moot now

File: pytato/transform.py [hunk 7]
```diff
@@ -994,16 +1076,102 @@ def find_partitions(expr: Array, part_func: Callable[[Array], Hashable]) ->\
...
+                tags=expr.tags)
+
+    def map_distributed_send(self, expr: DistributedSend) -> DistributedSend:
+        result = super().map_distributed_send(expr)
+
+        new_name = self.name_generator()
+        self.part_output_name_to_send_node[new_name] = result
+
+        # FIXME: correct? (used to be 'return result')
+        return expr.data  # type: ignore
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/transform.py]: done

**Comment by matthiasdiener** [comment]
partitioning:
image

**Comment by matthiasdiener** [comment]
With the recent updates, the partitioned code looks like this:

image


It currently fails with:

```
  File "examples/distributed.py", line 111, in main
    prg_per_partition = generate_code_for_partitions(parts)
  File "/Users/mdiener/Work/emirge/pytato/pytato/partition.py", line 328, in generate_code_for_partitions
    prg_per_partition[pid] = generate_loopy(d)
  File "/Users/mdiener/Work/emirge/pytato/pytato/target/loopy/codegen.py", line 910, in generate_loopy
    insn_id = add_store(name, expr, cg_mapper(expr, state), state, cg_mapper)
  File "/Users/mdiener/Work/emirge/pytato/pytato/transform.py", line 136, in __call__
    return self.rec(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/pytato/pytato/transform.py", line 132, in rec
    return method(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/pytato/pytato/target/loopy/codegen.py", line 376, in map_index_lambda
    loopy_expr = self.exprgen_mapper(expr.expr, prstnt_ctx, local_ctx)
  File "/Users/mdiener/Work/emirge/pytato/pytato/target/loopy/codegen.py", line 568, in __call__
    return self.rec(expr, prstnt_ctx, local_ctx)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 137, in __call__
    return method(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 422, in map_sum
    tuple(self.rec(child, *args, **kwargs) for child in expr.children)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 422, in <genexpr>
    tuple(self.rec(child, *args, **kwargs) for child in expr.children)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 137, in __call__
    return method(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/pytato/pytato/target/loopy/codegen.py", line 577, in map_subscript
    return result.to_loopy_expression(self.rec(expr.index, prstnt_ctx,
  File "/Users/mdiener/Work/emirge/pytato/pytato/target/loopy/codegen.py", line 217, in to_loopy_expression
    assert len(indices) == self.num_indices
AssertionError
```

@inducer

File: pytato/visualization.py [hunk 8]
```diff
@@ -320,6 +320,7 @@ def get_dot_graph_from_partitions(parts: CodePartitions) -> str:
                         emitted_placeholders.add(array)
 
                         # Emit cross-partition edges
+                        assert array.name
```
Comment (inducer) [inline pytato/visualization.py:324]: Shouldn't be necessary?

Comment (matthiasdiener) [reply pytato/visualization.py]: removed

**Comment by inducer** [comment]
The distributed example runs now. Important next steps would be to have actual tests of the distributed functionality, and to merge #208 once that lands.

File: test/test_distributed.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    axis_len = 4
+
+    ntests = 1
+    for i in range(ntests):
+        print(i)
+        seed = 120 + i
+        rdagc_pt = RandomDAGContext(np.random.default_rng(seed=seed),
+                axis_len=axis_len, use_numpy=False)
+
+        x = make_random_dag(rdagc_pt)
```
Comment (inducer) [inline test/test_distributed.py:89]: Could drill open `make_random_dag` to allow adding self-send/self-receive pairs.

File: pytato/distributed.py [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+
+    .. attribute:: dest_rank
+
+        An :class:`int`. The rank to which :attr:`data` is to be sent.
+
+    .. attribute:: comm_tag
+
+        A hashable, picklable object to serve as a 'tag' for the communication.
+        Only a :class:`DistributedRecv` with the same tag will be able to
+        receive the data being sent here.
```
Comment (inducer) [inline pytato/distributed.py:82]: Currently assuming integer. Should allow gathering/creating a mapping in comm info gathering.

File: test/test_distributed.py [hunk 11]
```diff
@@ -70,3 +69,56 @@ def test_distributed_execution_basic(ctx_factory):
...
+
+        halo = staple_distributed_send(x, dest_rank=(rank-1) % size, comm_tag=42,
+            stapled_to=make_distributed_recv(
+                src_rank=(rank+1) % size, comm_tag=42, shape=(4, 4), dtype=float))
```
Comment (inducer) [inline test/test_distributed.py:93]: `shape` not guaranteed to be right here.

File: test/test_distributed.py [hunk 12]
```diff
@@ -70,3 +69,56 @@ def test_distributed_execution_basic(ctx_factory):
...
+
+        y = halo * 42
+
+        res = staple_distributed_send(y, dest_rank=(rank-1) % size, comm_tag=42,
+            stapled_to=make_distributed_recv(
+                src_rank=(rank+1) % size, comm_tag=42, shape=(4, 4), dtype=float))
```
Comment (inducer) [inline test/test_distributed.py]: `shape` not guaranteed to be right here.

File: test/test_distributed.py [hunk 13]
```diff
@@ -70,3 +69,56 @@ def test_distributed_execution_basic(ctx_factory):
...
+
+        ctx = cl.create_some_context()
+        queue = cl.CommandQueue(ctx)
+
+        context = execute_partition_distributed(distributed_parts, prg_per_partition,
+                                                queue, comm)
+
+        final_res = [context[k] for k in dict_named_arys.keys()]
+
+        assert final_res
```
Comment (inducer) [inline test/test_distributed.py:116]: Compare to *something*?

File: pytato/distributed.py (hunk 10 above)
Comment (matthiasdiener) [reply pytato/distributed.py:88]: Should we postpone implementing this? This seems more like an add-on feature...

File: pytato/distributed.py [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+
+
+def _mpi_send(comm: Any, send_node: DistributedSend,
+             data: np.ndarray[Any, Any]) -> Any:
+    # Must use-non-blocking send, as blocking send may wait for a corresponding
+    # receive to be posted (but if sending to self, this may only occur later).
+    return comm.Isend(data, dest=send_node.dest_rank, tag=send_node.comm_tag)
+
+
+def execute_partition_distributed(
```
Comment (inducer) [inline pytato/distributed.py:436]: Maybe `execute_distributed_partition` would be better? (Also for other names in this module.)

Comment (inducer) [reply pytato/distributed.py]: Done in 9fbabc1.

File: pytato/distributed.py [hunk 15]
```diff
@@ -525,10 +525,17 @@ def execute_distributed_partition(
...
-    recv_names_tup, recv_requests_tup, recv_buffers_tup = zip(*[
-            (name,) + _post_receive(mpi_communicator, recv)
-            for part in partition.parts.values()
-            for name, recv in part.input_name_to_recv_node.items()])
+    try:
+        recv_names_tup, recv_requests_tup, recv_buffers_tup = zip(*[
+                (name,) + _post_receive(mpi_communicator, recv)
+                for part in partition.parts.values()
+                for name, recv in part.input_name_to_recv_node.items()])
+    except ValueError:
```
Comment (inducer) [inline pytato/distributed.py:562]: As error handling goes, this is too broad. There are probably hundreds of things in here that could cause a `ValueError`.

File: examples/distributed.py [hunk 16]
```diff
@@ -50,7 +50,7 @@ def main():
     context = execute_distributed_partition(distributed_parts, prg_per_partition,
                                              queue, comm)
 
-    final_res = [context[k] for k in outputs.keys()]
+    final_res = [context[k] for k in outputs.keys()][0].get(queue)
```
Comment (inducer) [inline examples/distributed.py:53]: What's the `[0]` doing? That looks like it's masking a bug elsewhere.

File: test/test_distributed.py [hunk 17]
```diff
@@ -65,9 +65,9 @@ def test_distributed_execution_basic(ctx_factory):
     context = execute_distributed_partition(distributed_parts, prg_per_partition,
                                              queue, comm)
 
-    final_res = [context[k] for k in outputs.keys()]
+    final_res = [context[k] for k in outputs.keys()][0].get(queue)
```
Comment (inducer) [inline test/test_distributed.py:68]: Also here: What's the `[0]` good for?

File: pytato/transform.py [hunk 18]
```diff
@@ -1388,10 +1388,15 @@ def map_loopy_call(self, expr: LoopyCall) -> LoopyCall:
...
-            self, expr: DistributedSendRefHolder, *args: Any) -> Any:
-        pass
+            self, expr: DistributedSendRefHolder, *args: Any) -> \
+                DistributedSendRefHolder:
+        from pytato.distributed import DistributedSendRefHolder
+        return DistributedSendRefHolder(
+            send=self.handle_edge(expr, expr.send.data),
+            passthrough_data=self.handle_edge(expr, expr.passthrough_data),
+            tags=expr.tags
+        )
```
Comment (inducer) [inline pytato/transform.py:1399]: How come this came back? In my view, this shouldn't be here.

File: pytato/distributed.py [hunk 19]
```diff
@@ -368,8 +393,9 @@ def map_distributed_send(self, expr: DistributedSend) -> DistributedSend:
         return new_send
 
 
-def _gather_distributed_comm_info(partition: GraphPartition) -> \
-        DistributedGraphPartition:
+def _gather_distributed_comm_info(partition: GraphPartition,
+        sends: Dict[PartId, List[DistributedSend]]) -> \
```
Comment (inducer) [inline pytato/distributed.py:397]: Improve argument name: `pid_to_distributed_sends` maybe?

File: pytato/transform.py (hunk 18 above)
Comment (matthiasdiener) [reply pytato/transform.py:1437]: I'm not sure what you mean here, could you clarify please?

File: pytato/distributed.py (hunk 15 above)
Comment (matthiasdiener) [reply pytato/distributed.py]: Does  b281b90 look better?

File: examples/distributed.py (hunk 16 above)
Comment (matthiasdiener) [reply examples/distributed.py]: f6ea090 simplified this.

File: test/test_distributed.py (hunk 17 above)
Comment (matthiasdiener) [reply test/test_distributed.py:68]: See f6ea090.

File: pytato/distributed.py (hunk 19 above)
Comment (matthiasdiener) [reply pytato/distributed.py]: Fixed in 9401b2e

File: pytato/distributed.py (hunk 15 above)
Comment (inducer) [reply pytato/distributed.py]: That's a (fairly big) merge commit. Did you make functional changes in that?

Comment (inducer) [reply pytato/distributed.py]: (Also, this file is unchanged in that commit.)

Comment (matthiasdiener) [reply pytato/distributed.py]: Whoops, this should have been b32abe13255d35b49dc01becc12b530bb5cd9d83

File: test/test_distributed.py (hunk 17 above)
Comment (inducer) [reply test/test_distributed.py:68]: Still there?

File: test/test_distributed.py (hunk 13 above)
Comment (matthiasdiener) [reply test/test_distributed.py]: Done in f3c9bc1

File: test/test_distributed.py (hunk 17 above)
Comment (matthiasdiener) [reply test/test_distributed.py:68]: Fixed in f3c9bc1

File: pytato/tags.py [hunk 20]
```diff
@@ -1,4 +1,7 @@
 __doc__ = """
+.. currentmodule:: pytato.tags
+
+
```
Comment (matthiasdiener) [inline pytato/tags.py:4]: This was removed in #245.

**Comment by matthiasdiener** [comment]
I think this is ready for review @inducer.

File: pytato/distributed.py (hunk 10 above)
Comment (matthiasdiener) [reply pytato/distributed.py:109]: Added in https://github.com/inducer/pytato/pull/148/commits/fc5df91baa73a57267a5a665c2c2cd9bd2d15ebb

File: test/test_distributed.py (hunk 9 above)
Comment (matthiasdiener) [reply test/test_distributed.py]: Added in 009142f

File: test/test_distributed.py [hunk 21]
```diff
@@ -116,53 +116,26 @@ def test_distributed_execution_random_dag():
...
-    rank = comm.Get_rank()
-    size = comm.Get_size()
-
     from testlib import RandomDAGContext, make_random_dag
 
     axis_len = 4
-    comm_fake_prob = 500
 
-    gen_comm_called = False
+    have_comm = False
```
[the remark this answers is not in the document]
Comment (inducer) [reply test/test_distributed.py:127]: 19bb3e4

File: pytato/partition.py [hunk 22]
```diff
@@ -41,6 +41,7 @@
 __doc__ = """
 .. autoclass:: GraphPart
 .. autoclass:: GraphPartition
+.. autoclass:: _GraphPartitioner
```
[the remark this answers is not in the document]
Comment (inducer) [reply pytato/partition.py]: 7830af7f049ddb898edc698dbc739c3785706040

File: pytato/distributed.py [hunk 23]
```diff
@@ -258,8 +258,6 @@ class DistributedGraphPart(GraphPart):
     .. attribute:: output_name_to_send_node
     .. attribute:: distributed_sends
     """
-    # TODO Document these
```
[the remark this answers is not in the document]
Comment (inducer) [reply pytato/distributed.py]: :shrug:

**Pull request merged**